libyul/optimiser/StackCompressor.cpp

```diff
@@ -50,31 +50,41 @@ namespace
 /**
  * Class that discovers all variables that can be fully eliminated by rematerialization,
  * and the corresponding approximate costs.
+ *
+ * Prerequisite: Disambiguator, Function Grouper
  */
 class RematCandidateSelector: public DataFlowAnalyzer
 {
 public:
 	explicit RematCandidateSelector(Dialect const& _dialect): DataFlowAnalyzer(_dialect) {}
 
-	/// @returns a map from rematerialisation costs to a vector of variables to rematerialise
+	/// @returns a map from function name to rematerialisation costs to a vector of variables to rematerialise
 	/// and variables that occur in their expression.
 	/// While the map is sorted by cost, the contained vectors are sorted by the order of occurrence.
-	map<size_t, vector<tuple<YulString, set<YulString>>>> candidates()
+	map<YulString, map<size_t, vector<tuple<YulString, set<YulString>>>>> candidates()
 	{
-		map<size_t, vector<tuple<YulString, set<YulString>>>> cand;
-		for (auto const& candidate: m_candidates)
+		map<YulString, map<size_t, vector<tuple<YulString, set<YulString>>>>> cand;
+		for (auto const& [functionName, candidate]: m_candidates)
 		{
 			if (size_t const* cost = util::valueOrNullptr(m_expressionCodeCost, candidate))
 			{
 				size_t numRef = m_numReferences[candidate];
 				set<YulString> const* ref = references(candidate);
```

Comment:
Oh wow! This uses the number of references to the candidate after the whole thing has run.
It should be zero at that point.

Reply:
Ok, the only influence this has is that we stop looking for candidates in the non-optimized codegen as soon as we would want to eliminate a variable that references one that has already been chosen for elimination.

So in the worst case, we eliminate more variables than we actually should.